src/rust/engine/rule_graph/src/lib.rs

```diff
@@ -307,6 +307,34 @@ impl<R: Rule> RuleGraph<R> {
             .collect()
     }
 
+    ///
+    /// Returns a mapping from a `Rule` to the dependencies it has that are themselves satisfied by
+    /// `Rule`s.
+    ///
+    #[allow(clippy::type_complexity)]
+    pub fn rule_dependencies(&self) -> HashMap<&R, Vec<(&DependencyKey<R::TypeId>, &R)>> {
+        let mut result = HashMap::default();
+        for (entry, rule_edges) in &self.rule_dependency_edges {
+            let EntryWithDeps::Rule(RuleEntry { rule, .. }) = entry.as_ref() else {
+                continue;
+            };
+
+            let mut function_satisfied_gets = Vec::new();
+            for (dependency, source) in &rule_edges.dependencies {
+                let Entry::WithDeps(entry_with_deps) = source.as_ref() else {
+                    continue;
+                };
+                let EntryWithDeps::Rule(RuleEntry { rule, .. }) = entry_with_deps.as_ref() else {
+                    continue;
+                };
+
+                function_satisfied_gets.push((dependency, rule));
```

Comment:
This local `rule` variable shadows the outer one, right?
Reading the code this is a bit confusing, as it's easy to miss (at least for my novice Rust eyes), and it made me think the `result.insert(rule, function_satisfied_gets)` uses the same `rule` as here, until I spotted this shadowing...

I guess the name comes from the pattern matching, but is there no way to use pattern matching, but capture the value in a customized variable name?